module/zfs/zfs_fm.c

```diff
@@ -904,19 +904,26 @@ zfs_ereport_post_checksum(spa_t *spa, vdev_t *vd,
 #endif
 }
 
-static void
-zfs_post_common(spa_t *spa, vdev_t *vd, const char *type, const char *name,
-    nvlist_t *aux)
+/*
+ * The 'sysevent.fs.zfs.*' events are signals posted to notify user space of
+ * change in the pool.  All sysevents are listed in sys/sysevent/eventdefs.h
+ * and are designed to be consumed by the ZFS Event Daemon (ZED).  For
+ * additional details refer to the zed(8) man page.
+ */
+sysevent_t *
+zfs_create_sysevent_impl(spa_t *spa, vdev_t *vd, const char *type,
+    const char *name, nvlist_t *aux)
```

Comment:
It would be nice if we could largely confine the sysevent handling to `spa.c` as is done in OpenZFS.  All we should really need in `zfs_fm.c` is your new `zfs_post_sysevent()` function and access to the packing logic in `zfs_post_common()`.  You've already refactored this logic out to return a `sysevent_t`, you'll be almost there if instead you return the `nvlist_t` and rename the function `zfs_post_create()`.  Then  `spa_event_create()` can call the new `zfs_post_create()` and handle the `sysevent_t` wrapper, and `zfs_post_common()` can also call `zfs_post_create` prior to `zfs_zevent_post()`.  That should largely get all of this code aligned with OpenZFS as much as is reasonable.